js/private/js_library_helpers.bzl

```
@@ -69,104 +69,91 @@ def gather_npm_linked_packages(srcs, deps):
         deps: dep targets; these typically come from the `deps` attribute of a rule
 
     Returns:
-        A `struct(direct, transitive)` of direct and transitive npm linked packages gathered
+        A `struct(direct, direct_files, transitive, transitive_files)` of direct and transitive npm linked packages & underlying files gathered
     """
 
     # npm_linked_packages
     npm_linked_packages = [
-        item
+        target[JsInfo].npm_linked_packages
         for target in srcs
-        if JsInfo in target and hasattr(target[JsInfo], "npm_linked_packages")
-        for item in target[JsInfo].npm_linked_packages
+        if JsInfo in target
+    ]
+
+    # npm_linked_package_files
+    npm_linked_package_files = [
+        target[JsInfo].npm_linked_package_files
+        for target in srcs
+        if JsInfo in target
     ]
-    npm_linked_packages = [package for package in npm_linked_packages if package]
 
     # transitive_npm_linked_packages
-    transitive_npm_linked_packages = npm_linked_packages + [
-        item
+    transitive_npm_linked_packages = depset([], transitive = npm_linked_packages + [
+        target[JsInfo].transitive_npm_linked_packages
         for target in srcs + deps
-        if JsInfo in target and hasattr(target[JsInfo], "transitive_npm_linked_packages")
-        for item in target[JsInfo].transitive_npm_linked_packages
-    ]
-    transitive_npm_linked_packages = [package for package in transitive_npm_linked_packages if package]
+        if JsInfo in target
+    ])
+
+    # transitive_npm_linked_package_files
+    transitive_npm_linked_package_files = depset([], transitive = npm_linked_package_files + [
+        target[JsInfo].transitive_npm_linked_package_files
+        for target in srcs + deps
+        if JsInfo in target
+    ])
 
     return struct(
-        direct = npm_linked_packages,
+        direct = depset([], transitive = npm_linked_packages),
+        direct_files = depset([], transitive = npm_linked_package_files),
         transitive = transitive_npm_linked_packages,
+        transitive_files = transitive_npm_linked_package_files,
     )
 
-def gather_npm_package_stores(targets):
+def gather_npm_package_store_deps(targets):
     """Gathers NpmPackageStoreInfo providers from the list of targets
 
     Args:
         targets: the list of targets to gather from
 
     Returns:
-        A `struct(direct, transitive)` of direct and transitive npm package stores gathered
+        Depset of npm package stores gathered
```

```suggestion
        A `depset` of npm package stores gathered
```
Similar to the last suggestion. Just seems like when referring to a type we shouldn't change the case of anything...